### PR TITLE
Implement the rest of Upgrade, UpgradeTemplate and UpgradeCenter.

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/system/upgrade.cpp [hunk 1]
```diff
@@ -102,3 +117,140 @@ bool UpgradeCenter::Can_Afford_Upgrade(Player *player, const UpgradeTemplate *up
...
+    m_buttonImage(nullptr)
+{
+}
+
+int UpgradeTemplate::Calc_Time_To_Build(Player *player) const
+{
+#ifdef GAME_DEBUG_STRUCTS
+    if (player->Is_Instant_Build()) {
+        return 1;
+    } else
```
Comment (xezon) [inline src/game/common/system/upgrade.cpp:212]: Can remove the `else` to simplify.

File: src/game/common/system/upgrade.cpp [hunk 2]
```diff
@@ -102,3 +117,140 @@ bool UpgradeCenter::Can_Afford_Upgrade(Player *player, const UpgradeTemplate *up
...
+#endif
+    {
+        return m_buildTime * 30.0f;
```
Comment (xezon) [inline src/game/common/system/upgrade.cpp:215]: This returns `float` on `int`. As designed?

File: src/game/common/system/upgrade.cpp [hunk 3]
```diff
@@ -33,19 +33,23 @@ UpgradeCenter::~UpgradeCenter()
...
 {
-#ifdef GAME_DLL
-    Call_Method<void, UpgradeCenter>(PICK_ADDRESS(0x004AF190, 0x00728DBD), this);
-#endif
+    if (g_theMappedImageCollection != nullptr && !m_buttonImagesCached) {
+        for (UpgradeTemplate *upgrade = m_upgradeList; upgrade != nullptr; upgrade = upgrade->Friend_Get_Next()) {
+            upgrade->Cache_Button_Image();
+        }
+
+        m_buttonImagesCached = true;
```
Comment (xezon) [inline src/game/common/system/upgrade.cpp:48]: This is surprising. Should the caching not happen in the Init?

**Review by xezon — APPROVED** [review]
Tested. Upgrades work normally. Veterancy also works normally.

File: src/game/common/system/upgrade.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/common/system/upgrade.cpp:48]: This is correct to the original.

File: src/game/common/system/upgrade.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/common/system/upgrade.cpp:212]: This is correct to the original.

File: src/game/common/system/upgrade.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/common/system/upgrade.cpp:215]: This is correct to the original.

**Pull request merged**